validator/iam_allow_ban_roles.rego

```diff
@@ -0,0 +1,59 @@
+#
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package templates.gcp.GCPIAMAllowBanRolesConstraintV1
+
+import data.validator.gcp.lib as lib
+
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	lib.get_constraint_params(constraint, params)
+	asset := input.asset
+
+	binding := asset.iam_policy.bindings[_]
+	role := binding.role
+
+	matches_found = {r | r := role; glob.match(params.roles[_], [], r)}
+
+	mode := lib.get_default(params, "mode", "allow")
+
+	target_match_count(mode, desired_count)
```

Comment:
I'm a little new to rego, so I'm not sure what this is doing here.  [The rego docs](https://www.openpolicyagent.org/docs/latest/how-do-i-write-policies/#functions) don't seem to define what happens if a function is given more arguments than it's defined with, so it's not clear to me what the value of "desired_count" will evaluate to here.  What does this value represent?

Reply:
I copied this implementation from "iam_allowed_bindings.rego". I think `desired_count` stores the returned value of `target_match_count`. If `mode == "ban"`, `desired_count` evaluates to 0, otherwise 1.

Reply:
This code is equivalent:

```
desired_count := target_match_count(mode)
```

Reply:
Code updated to use syntax `desired_count := target_match_count(mode)` as it is more readable than the original version.